codalab/codabench
# Competition Whitelist Emails feature

# @ mention of reviewers
@Didayolo 


# A brief description of the purpose of the changes contained in this PR.
Now, users can add a list of emails to their competition named `Whitelist Emails`. If this competition is not open i.e. participants are approved by an organizer, then any user whose email is added in whitelist_emails will not require an approval i.e. on register button click, this user will be approved automatically.


# Process

#### Add whitelist emails list
Users can add the whitelist emails list here: Competition -> Edit -> Participation -> Scroll down to white List -> Follow the written instructions

<img width="940" alt="Screenshot 2023-11-12 at 9 17 24 PM" src="https://github.com/codalab/codabench/assets/13259262/c9358676-fd5b-4203-8b65-8e3791910114">


#### Error handling
If a user enters wrong emails, emails are validated and errors are shown
<img width="905" alt="Screenshot 2023-11-12 at 9 18 46 PM" src="https://github.com/codalab/codabench/assets/13259262/ac680dee-f9a0-42ab-b845-a65f73414b62">


# How to test this feature
1. Add 1 email addresses to the white list like the way described above
2. Make sure to disable auto approve participants
<img width="290" alt="Screenshot 2023-11-12 at 9 19 49 PM" src="https://github.com/codalab/codabench/assets/13259262/8720b600-3fbc-445a-8419-ac703dc2f69e">

3. Login with the email added in the whitelist, go to the competition registration page and click register, you will see that you are auto approved without the need of an admin to approve you.

# Issues this PR resolves
- #1064 



# Checklist
- [x] I'm proud of my work
- [x] Code review by reviewer
- [x] CircleCi tests are passing
